### PR TITLE
Add lint when no doc is present at the crate-level

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Follow-up of #66267.

r? @kinnison

### CODE REVIEW

File: src/librustdoc/core.rs [hunk 1]
```diff
@@ -415,6 +419,23 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
 
             let mut krate = clean::krate(&mut ctxt);
 
+            if let Some(ref m) = krate.module {
+                match m.doc_value() {
+                    None | Some("") => {
+                        let mut diag = tcx.struct_lint_node(
+                            rustc_lint::builtin::NO_CRATE_LEVEL_DOC,
+                            ctxt.as_local_hir_id(m.def_id).unwrap(),
+                            "No documentation found on this crate top module.\n\n\
```
Comment (jonas-schievink) [inline src/librustdoc/core.rs]:
```suggestion
                            "no documentation found on this crate top module.\n\n\
```

File: src/librustdoc/core.rs [hunk 2]
```diff
@@ -415,6 +419,23 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
...
+                             Maybe you could be interested into looking at this documentation:\n\
```
Comment (jonas-schievink) [inline src/librustdoc/core.rs]: This should be a `note` or a `help`, otherwise it messes up the output

**Review by jonas-schievink — COMMENTED** [review]
Not sure what the motivation for this is given that we already have a missing docs lint

File: src/librustdoc/core.rs [hunk 3]
```diff
@@ -415,6 +419,23 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
 
             let mut krate = clean::krate(&mut ctxt);
 
+            if let Some(ref m) = krate.module {
+                match m.doc_value() {
+                    None | Some("") => {
```
Comment (Centril) [inline src/librustdoc/core.rs]:
```suggestion
                if let None | Some("") = m.doc_value() {
```

**Comment by GuillaumeGomez** [comment]
@jonas-schievink I'd like to have this lint activated at the warning level by default once we all agree on it (and once the rustdoc book page is stable too). The idea is to provide some help by default for people that don't know how to write documentation.

Hmmm... Maybe I should have written that in the PR description...

File: src/librustdoc/core.rs (hunk 1 above)
Comment (kinnison) [inline src/librustdoc/core.rs]: Not sure how we really talk about the top level of a crate, is it a module per-se?

If so, perhaps:

```suggestion
                            "no documentation found for this crate's top-level module.\n\n\
```

File: src/librustdoc/core.rs (hunk 2 above)
Comment (kinnison) [inline src/librustdoc/core.rs]:
```suggestion
                             The following guide may be of use:\n\
```

**Review by kinnison — COMMENTED** [review]
A couple of potential wording tweaks in addition to the other comments.

File: src/librustc/lint/builtin.rs [hunk 4]
```diff
@@ -337,6 +337,12 @@ declare_lint! {
     "failures in resolving intra-doc link targets"
 }
 
+declare_lint! {
+    pub NO_CRATE_LEVEL_DOC,
```
Comment (phansch) [inline src/librustc/lint/builtin.rs]: I think `MISSING_CRATE_LEVEL_DOC` would be a be a bit better in communicating what's wrong. Also it reads more fluidly when used with `allow/warn/deny` (see [RFC-0344#lints](https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints))

Comment (GuillaumeGomez) [reply src/librustc/lint/builtin.rs]: :+1:

**Comment by joelpalmer** [comment]
Ping from Triage: any updates @GuillaumeGomez?

**Comment by GuillaumeGomez** [comment]
I need to update the PR. Lack of time lately...

**Comment by JohnTitor** [comment]
Ping from triage: @GuillaumeGomez any updates on this? This PR is idle for about two months.

**Comment by GuillaumeGomez** [comment]
Forgot about it... I'll try to get back to it "soon".

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by petrochenkov** [comment]
This PR is still missing motivation.

If the lint is not enabled by default, then we already have `missing_docs` that can be enabled.
If the lint is enabled by default in `rustdoc` runs, then we already have `missing_docs` that can be reported by `rustdoc` as well, it should have all the necessary information.

**Comment by GuillaumeGomez** [comment]
The motivation is providing help for a specific kind of documentation. It'll also be set as warning by default.

As a sidenote, need to check what failed the tests...

**Comment by GuillaumeGomez** [comment]
Oh, the lint sentence changed, nice.

**Comment by GuillaumeGomez** [comment]
Ping @kinnison

**Comment by kinnison** [comment]
All in all I'm OK with the current state of the change. :+1:

My only concern is whether @petrochenkov 's worries remain valid or if they have been assuaged by Guillaume's explanations.

If @petrochenkov is OK with things then I'm okay to r+ this (or someone else can r=kinnison).

**Comment by petrochenkov** [comment]
I still don't understand why you can't report existing `missing_docs` from rustdoc instead, https://github.com/rust-lang/rust/pull/66938#issuecomment-584684822 didn't answer any of my questions.

**Comment by GuillaumeGomez** [comment]
I guess I badly explained... Let's try again! :)

So, the idea behind this lint is to help newcomers using rustdoc when they haven't documenting much (or nothing). This lint will be enabled at a warning level by default so it'll appear every time. It is **different** from `missing_docs` lint in this regard because this new one isn't intend to look for missing docs but to give a starting point for people who don't know where to start to document their code.

**Comment by joelpalmer** [comment]
Triaged

**Comment by Dylan-DPC** [comment]
@petrochenkov do you still have concerns with this PR?

**Comment by petrochenkov** [comment]
I still have the concerns about insufficient motivation, and I'm not entirely sure this can be enabled by default.

I think this needs a feedback from larger audience.
RFC process is perhaps too heavy-weight for this, but it would be adequate to have a post on internals explaining in detail what this lints want to achieve and why it is useful.

File: src/test/rustdoc-ui/no-crate-level-doc-lint.rs [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#![deny(missing_crate_level_doc)]
```
Comment (petrochenkov) [inline src/test/rustdoc-ui/no-crate-level-doc-lint.rs]:
```suggestion
#![deny(missing_crate_level_docs)]
```
or better
```suggestion
#![deny(missing_crate_docs)]
```
We have naming conventions for lints - https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints.

**Comment by GuillaumeGomez** [comment]
> I still have the concerns about insufficient motivation, and I'm not entirely sure this can be enabled by default.

It'll be visible only when running rustdoc, so unless you want to document your API, you'll never this warning.

> I think this needs a feedback from larger audience.
> RFC process is perhaps too heavy-weight for this, but it would be adequate to have a post on internals explaining in detail what this lints want to achieve and why it is useful and why it can be enabled by default.

If you see this warning, it means that you're trying to get your API documentation. At this point, if you don't know about **how** to document rust code, it'll come in handy. Otherwise, you can either ignore the lint or allow it. In any case, it won't impact much people who don't have the need for it.

**Comment by Dylan-DPC** [comment]
eh sorry forgot to refresh the tab :D 

@bors r-

**Comment by GuillaumeGomez** [comment]
> eh sorry forgot to refresh the tab :D

?

**Pull request merged**